library/core/src/array/eq.rs

```diff
@@ -0,0 +1,160 @@
+        self[..] == other[..]
+    }
+    #[inline]
+    fn ne(&self, other: &[B]) -> bool {
+        self[..] != other[..]
+    }
+}
+
+#[stable(feature = "rust1", since = "1.0.0")]
+impl<A, B, const N: usize> PartialEq<[A; N]> for [B]
+where
+    B: PartialEq<A>,
+{
+    #[inline]
+    fn eq(&self, other: &[A; N]) -> bool {
+        self[..] == other[..]
+    }
+    #[inline]
+    fn ne(&self, other: &[A; N]) -> bool {
+        self[..] != other[..]
+    }
+}
+
+#[stable(feature = "rust1", since = "1.0.0")]
+impl<A, B, const N: usize> PartialEq<&[B]> for [A; N]
+where
+    A: PartialEq<B>,
+{
+    #[inline]
+    fn eq(&self, other: &&[B]) -> bool {
+        self[..] == other[..]
+    }
+    #[inline]
+    fn ne(&self, other: &&[B]) -> bool {
+        self[..] != other[..]
+    }
+}
+
+#[stable(feature = "rust1", since = "1.0.0")]
+impl<A, B, const N: usize> PartialEq<[A; N]> for &[B]
+where
+    B: PartialEq<A>,
+{
+    #[inline]
+    fn eq(&self, other: &[A; N]) -> bool {
+        self[..] == other[..]
+    }
+    #[inline]
+    fn ne(&self, other: &[A; N]) -> bool {
+        self[..] != other[..]
+    }
+}
+
+#[stable(feature = "rust1", since = "1.0.0")]
+impl<A, B, const N: usize> PartialEq<&mut [B]> for [A; N]
+where
+    A: PartialEq<B>,
+{
+    #[inline]
+    fn eq(&self, other: &&mut [B]) -> bool {
+        self[..] == other[..]
+    }
+    #[inline]
+    fn ne(&self, other: &&mut [B]) -> bool {
+        self[..] != other[..]
+    }
+}
+
+#[stable(feature = "rust1", since = "1.0.0")]
+impl<A, B, const N: usize> PartialEq<[A; N]> for &mut [B]
+where
+    B: PartialEq<A>,
+{
+    #[inline]
+    fn eq(&self, other: &[A; N]) -> bool {
+        self[..] == other[..]
+    }
+    #[inline]
+    fn ne(&self, other: &[A; N]) -> bool {
+        self[..] != other[..]
+    }
+}
+
+// NOTE: some less important impls are omitted to reduce code bloat
+// __impl_slice_eq2! { [A; $N], &'b [B; $N] }
+// __impl_slice_eq2! { [A; $N], &'b mut [B; $N] }
+
+#[stable(feature = "rust1", since = "1.0.0")]
+impl<T: Eq, const N: usize> Eq for [T; N] {}
+
+trait SpecArrayEq<Other, const N: usize>: Sized {
+    fn spec_eq(a: &[Self; N], b: &[Other; N]) -> bool;
+    fn spec_ne(a: &[Self; N], b: &[Other; N]) -> bool;
+}
+
+impl<T: PartialEq<Other>, Other, const N: usize> SpecArrayEq<Other, N> for T {
+    default fn spec_eq(a: &[Self; N], b: &[Other; N]) -> bool {
+        a[..] == b[..]
+    }
+    default fn spec_ne(a: &[Self; N], b: &[Other; N]) -> bool {
+        a[..] != b[..]
+    }
+}
+
+impl<T: PartialEq<U> + IsRawEqComparable<U>, U, const N: usize> SpecArrayEq<U, N> for T {
+    #[cfg(bootstrap)]
+    fn spec_eq(a: &[T; N], b: &[U; N]) -> bool {
+        a[..] == b[..]
+    }
+    #[cfg(not(bootstrap))]
+    fn spec_eq(a: &[T; N], b: &[U; N]) -> bool {
+        // SAFETY: This is why `IsRawEqComparable` is an `unsafe trait`.
+        unsafe {
+            let b = &*b.as_ptr().cast::<[T; N]>();
+            crate::intrinsics::raw_eq(a, b)
```

Comment:
Could we write the intrinsics expanded logic as ru if not, we couldat least do the length check here, rig he?  st code here instead of doing various codegen things in the intrinsic expansion?

Reply:
I don't know any way to write rust code that produces things like `icmp eq i96` and `icmp eq i256`.  I could write rust code that manually expands it out to loading the pieces as `u64` or something, but having a loop generating a bunch of that manually makes it much harder for LLVM's optimizer to understand what's going on -- when it's a single `icmp`, things like `a == [0; N]` collapse beautifully.

I actually tried putting the length check in here first.  That worked great in the optimized code, but it was horrific for the `opt-level=0` codegen of `array_eq_long` -- the `icmp eq i19744` was unreachable, but it didn't know that and proceeded to emit over 300 qword comparisons into the assembly output.

I'd definitely *rather* have the length check in the library -- it seems reasonable that different places might want different thresholds, and I'd prefer to call `memcmp` from there instead of the intrinsic -- so if you have a good solution for that I'd be happy to update things.

Reply:
Oof yea.. debug mode :/

We could allow trivial const prop for dead arm elimination even in debug mode. Like... not actually change anything except to replace the unreachable arms with zero statements and an unreachable intrinsic.

I'm assuming you meant to keep the intrinsic for the short length arm, which seems fine to me

Reply:
Idea: make the intrinsic generate an undef when called with lengths over 4xWord and document that in the intrinsic docs. Horribly unsafe, but so are most intrinsics

Reply:
I think that'd work, but I really don't feel like it'd be a good idea to spread knowledge of the poorly-justified threshold out between all four of core/codegen_llvm/codegen_cranelift/miri, even if it's documented.  I'd like it to be easy for someone to test things like "it should be `>=`, not `>`", for example.

Having slept on this, I think I actually *like* leaving it up to codegen to decide where the threshold is.  That's basically what was happening with the `memcmp` call -- the target-dependent codegen decided whether to leave in the function call or to emit loads instead.  (Except that that was too late for the optimizer to be able to also remove the `alloca`s, leading to the unfortunate "pass the arguments in registers, then write them to stack, then read them from stack again, then look at them".)

It wouldn't seem unreasonable, for example, to do something like "well, if `sess.opts.optimize == OptLevel::Aggressive`, then allow up to 8×usize".  (Though I'm not sure that that *specific* change would be worth the testing implications of having codegen emit different IR at different optimization levels, but maybe if well-justified it could be worth it like running mir-opts can be.)

And from a semantic cleanliness perspective, I'd much rather the definition be "it compares them", than "it compares them, unless they're bigger than 4×usize, in which case it aborts".  The MIRI implementation is nice and clean right now, for example, and adding something like a threshold into it makes it feel worse to me.